Fix bug 1426676: Support for WebExtensions' JSON format

mathjazz opened the pull request:
We also store `placeholder` data, which includes examples. Might come in handy in the future.

This is not urgent and can wait until @phlax is back.

Pike commented:
Wondering, are there ways to test this code?

Pike commented:
> We also store placeholder data

... couldn't find that in the code? Maybe I overlooked.

I'd also like to see us maintaining VCS attribution in the serialized file. AFAICT, json.dumps can serialize in whichever order it wants, and indention would also not be maintained.

mathjazz commented:
Yup, I should add tests here:
https://github.com/mathjazz/pontoon/tree/e735592e7f0c9768336a7979506a253d0cd2c0ae/pontoon/sync/tests/formats

Looks like I forgot to push the commit which imports placeholder data. Fixing.

Any suggestions on how to keep the output as close to the existing localized file as possible?

jotes commented on pontoon/sync/formats/json_extensions.py:
```diff
@@ -0,0 +1,138 @@
+            else:
+                raise
+
+        for order, unit in enumerate(self.json_file.items()):
```
`for order, (key, data) in enumerate(`

jotes commented on pontoon/sync/formats/json_extensions.py:
```diff
@@ -0,0 +1,138 @@
+                            .format(self.path))
+
+        with codecs.open(self.source_resource.path, 'r', 'utf-8') as resource:
+            json_file = json.loads(resource.read())
```
`json.load(resource)` - `.load` is able to read contents from file-like objects.

jotes commented on pontoon/base/models.py:
```diff
@@ -1665,15 +1665,16 @@ class Resource(models.Model):
+        ('ftl', 'ftl'),
         ('inc', 'inc'),
         ('ini', 'ini'),
+        ('json', 'json'),
```
`.json` is the general format for various types of data. I'm not sure how the layout of the future repositories will look like.
Maybe we should restrict our parsers only to  parse only files named `messages.json` and additionaly use http://json-schema.org/ to validate if the schema is correct?

phlax replied:
ive been using json-schema and it rocks - seems like a good way to validate files are actually messages

mathjazz replied:
Thanks, I'll look into that.

phlax commented on pontoon/sync/formats/json_extensions.py:
```diff
@@ -0,0 +1,138 @@
+    def fuzzy(self):
+        return False
+
+    @fuzzy.setter
```
mozilla `.lang` files also have a fuzzy logic i think - not to detract from the point about refactoring

mathjazz replied:
I'll look into refactoring the defaults. There's no fuzzy in .lang files.

phlax commented on pontoon/sync/formats/json_extensions.py:
```diff
@@ -0,0 +1,138 @@
+        try:
+            with codecs.open(path, 'r', 'utf-8') as resource:
+                content = resource.read()
+                self.json_file = json.loads(content)
```
unless there is some setting somewhere else this will use dictionaries for json objects - ie unordered

you can set it globally (which it might be, but i cant see it grepping), or in the call to `loads` - eg

```
json.loads(content, object_pairs_hook=OrderedDict)
```

phlax commented:
definitely could do with some tests - https://codecov.io/gh/mozilla/pontoon/pull/786/diff

mathjazz commented on pontoon/sync/formats/json_extensions.py:
```diff
@@ -0,0 +1,138 @@
+            with codecs.open(path, 'r', 'utf-8') as resource:
+                content = resource.read()
+                self.json_file = json.loads(content)
+        except IOError:
```
Turns out `json` raises `ValueError` and `simplejson` raised `JSONDecodeError`. Anyhow, I'll update.

mathjazz commented:
Given that this PR is now less of a priority than it was at the time of openning it, and it's not even clear when we'll start using it, I'd like to finalize it without impacting other file formats too much (e.g. fuzzy refactoring).

I've addressed most of the comments and added a frontend fix.

@phlax @jotes Let me know if this looks better now and then I'll add some tests.

phlax commented on pontoon/sync/formats/json_extensions.py:
```diff
@@ -0,0 +1,187 @@
+
+        try:
+            with codecs.open(path, 'r', 'utf-8') as resource:
+                self.json_file = json.load(resource)
```
i think it needs object_pairs_hook here too

phlax commented on pontoon/sync/formats/json_extensions.py:
```diff
@@ -0,0 +1,187 @@
+
+SCHEMA = {
+    "type": "object",
+    "patternProperties": {
```
im wondering if this would be simpler as just properties without the pattern match

mathjazz replied:
I'm not sure how to do that. The JSON file contains a [list of objects](https://github.com/mathjazz/pontoon-test/blob/master/en-US/example.json) where anything can be a key.

phlax replied:
just use `properties`

phlax replied:
aplogies - above comments should have said `additionalProperties` rather than `properties`

phlax commented on pontoon/sync/formats/json_extensions.py:
```diff
@@ -0,0 +1,187 @@
+                    }
+                }
+            },
+            "required": ["message", "description"]
```
this needs to be at the same level as the properties its constraining

mathjazz replied:
The nesting level seems to be fine?
http://json-schema.org/examples.html

It also breaks if I move it one level up.

phlax replied:
my bad - you're right

Pike commented on pontoon/sync/formats/json_extensions.py:
```diff
@@ -0,0 +1,179 @@
+        resource.
+        """
+        if not self.source_resource:
+            raise SyncError('Cannot save FTL resource {0}: No source resource given.'
```
Not FTL, but ... ?

Pike commented on pontoon/sync/formats/json_extensions.py:
```diff
@@ -0,0 +1,179 @@
+            try:
+                validate(json_file, SCHEMA)
+            except ValidationError as e:
+                raise ParseError(e)
```
Should we really refuse to sync if a translated json isn't valid? Or should we rather fix it up?

Pike commented on pontoon/sync/formats/json_extensions.py:
```diff
@@ -0,0 +1,179 @@
+            pass  # Already exists, phew!
+
+        with codecs.open(self.path, 'w+', 'utf-8') as f:
+            log.debug('Saved file: %s', self.path)
```
`Saved` is the wrong tense here, as above.

mathjazz commented:
@Pike @jotes I addressed the review comments and added tests. I also added support for marking up `$PLACEHOLDER$` as placeables and displaying placeholder examples in the metadata section (under the source string) in the editor.

I'd appreciate if you could have another look.

jotes commented on pontoon/base/static/js/translate.js:
```diff
@@ -589,10 +589,23 @@ var Pontoon = (function (my) {
+              examples.push('$' + this.toUpperCase() + '$: ' + example);
+            }
           });
+          self.appendMetaData('Placeholder Examples', examples.join(', '));
```
[nit/question] I'm not sure if `examples` will always contain escaped values (?). What do you think about `self.doNotRender` here?

mathjazz commented on pontoon/sync/formats/json_extensions.py:
```diff
@@ -0,0 +1,179 @@
+            if source_resource:
+                return
+            else:
+                raise ParseError(err)
```
Mostly for consistency with other formats, so it's easier to catch the exception:
https://github.com/mozilla/pontoon/blob/master/pontoon/sync/vcs/models.py#L349

mathjazz commented on pontoon/sync/formats/json_extensions.py:
```diff
@@ -0,0 +1,184 @@
+        try:
+            os.makedirs(os.path.dirname(self.path))
+        except OSError:
+            pass  # Already exists, phew!
```
We use the same block in `silme` and `FTL`, so I'll factor it out. Note that `os.path.isdir(self.path)` will always be false, because `self.path` is the path to a resource file.

mathjazz commented on pontoon/sync/formats/json_extensions.py:
```diff
@@ -0,0 +1,184 @@
+
+        with codecs.open(self.path, 'w+', 'utf-8') as f:
+            log.debug('Saving file: %s', self.path)
+            f.write(json.dumps(json_file, indent=4))
```
Again, this is consistent with `FTL` and `silme` formats, in which we also use the source file as the basis, remove untranslated and update translated strings, and then dump the tree into a localized file.

That sometimes causes an unnecessarily large diff, but fixing that problem will need to wait for a more extensive refactoring.

mathjazz commented:
I addressed your comments, @jotes. Thanks for your feedback!

For testing, you can use this:
https://github.com/mathjazz/pontoon-test
